### ISSUE TITLE
Callbacks on `save_post` with priority lower than 10 were ignored for the source blog

### ISSUE THREAD

**Issue by dnaber-de** [inpsyde/multilingual-press#96]
Here's the situation:

WordPress 4.1.1 (virtual localhost, recently updated from 3.8.5) with MLP 2.1.2 (pro, recently checked out the 2.1.2 tag coming from an outdated master branch). Two sites are set up (English and German) and are related to each other.

The English site has the ID 1, the German site has the ID 2.

When I create a post on the English site (`/wp-admin/post-new.php`) and save it, everything works as expected. As soon as I have a related translation (draft or publish) any callback on the action `save_post` with the priority 11 or lower (means number 11 or higher) never gets called in the context of the English (ID 1) site.

Here's a simple plugin which shows the behaviour (activated network wide):

``` php
<?php

/**
 * Plugin Name: MLP Debug
 * Version: 2015.03.20
 */

namespace MLPDebug;

add_action( 'wp_loaded', __NAMESPACE__ . '\init' );
function init() {

    add_action( 'save_post', __NAMESPACE__ . '\on_save_post_10', 10 );
    add_action( 'save_post', __NAMESPACE__ . '\on_save_post_11', 11 );

    add_action( 'mlp_before_post_synchronization', __NAMESPACE__ . '\on_mlp_before_post_synchronization' );
    add_action( 'mlp_after_post_synchronization', __NAMESPACE__ . '\on_mlp_after_post_synchronization' );
}

function on_save_post_11() {

    $msg = 'save_post_callback_11/blog/' . get_current_blog_id();
    trigger_error( $msg, E_USER_NOTICE );
}

function on_save_post_10() {

    $msg = 'save_post_callback_10/blog/' . get_current_blog_id();
    trigger_error( $msg, E_USER_NOTICE );
}

function on_mlp_before_post_synchronization() {

    $msg = 'mlp_before_post_synchronization_callback_10/blog/' . get_current_blog_id();
    trigger_error( $msg, E_USER_NOTICE );
}

function on_mlp_after_post_synchronization() {

    $msg = 'mlp_after_post_synchronization_callback_10/blog/' . get_current_blog_id();
    trigger_error( $msg, E_USER_NOTICE );
}

```

The following error message order I would expect (remember Site 1 is the »current« one):
- `mlp_before_post_synchronization_callback_10/blog/1` because the MLP `save_post` callback is registered before `MLPDebug\save_post_callback_10`. 
- `save_post_callback_10/blog/2` because mlp has switched the blog context already
- `save_post_callback_11/blog/2` of course
- `mlp_after_post_synchronization_callback_10/blog/` MLPs `save_post` callback is done
- `save_post_callback_10/blog/1` which is on the same priority stack as MLP but runs later
- `save_post_callback_11/blog/1` because it's the next lower priority in the list of filters

What I actually see is:

```
[Fri Mar 20 15:42:56.079963 2015] [:error] [pid 1532] [client 127.0.0.1:60890] PHP Notice:  mlp_before_post_synchronization_callback_10/blog/1 in wp-content/plugins/mlp-debug/mlp-debug.php on line 35
[Fri Mar 20 15:42:56.100159 2015] [:error] [pid 1532] [client 127.0.0.1:60890] PHP Notice:  save_post_callback_10/blog/2 in wp-content/plugins/mlp-debug/mlp-debug.php on line 29
[Fri Mar 20 15:42:56.100276 2015] [:error] [pid 1532] [client 127.0.0.1:60890] PHP Notice:  save_post_callback_11/blog/2 in wp-content/plugins/mlp-debug/mlp-debug.php on line 23
[Fri Mar 20 15:42:56.103903 2015] [:error] [pid 1532] [client 127.0.0.1:60890] PHP Notice:  mlp_after_post_synchronization_callback_10/blog/1 in wp-content/plugins/mlp-debug/mlp-debug.php on line 41
[Fri Mar 20 15:42:56.107458 2015] [:error] [pid 1532] [client 127.0.0.1:60890] PHP Notice:  save_post_callback_10/blog/1 in wp-content/plugins/mlp-debug/mlp-debug.php on line 29
```

In other words: `MLPDebug\on_save_post_11` never gets called in a context of site 1.

**Comment by dnaber-de** [comment]
A simplified variant of the plugin, thanks to @toscho 

```php
<?php

/**
 * Plugin Name: MLP Debug
 * Version: 2015.03.20
 */

namespace MLPDebug;

add_action( 'save_post', __NAMESPACE__ . '\log', 10 );
add_action( 'save_post', __NAMESPACE__ . '\log', 11 );
add_action( 'mlp_before_post_synchronization', __NAMESPACE__ . '\log' );
add_action( 'mlp_after_post_synchronization', __NAMESPACE__ . '\log' );

function log( $msg ) {
	global $wp_filter;

	$file     = WP_CONTENT_DIR . '/debug.log';
	$hook     = current_filter();
	$priority = key( $wp_filter[ $hook ] );
	$site     = get_current_blog_id();
	$msg      = "$hook/$priority/$site\n";

	file_put_contents( $file, $msg, FILE_APPEND | LOCK_EX );
}
```
Expected output:
```
save_post/10/1
mlp_before_post_synchronization/10/1
save_post/10/2
save_post/11/2
mlp_after_post_synchronization/10/1
save_post/10/2
```

Actual output:
```
save_post/10/1
mlp_before_post_synchronization/10/1
save_post/10/2
save_post/11/2
mlp_after_post_synchronization/10/1
```

**Comment by christianwach** [comment]
@dnaber-de I've noticed the same - that the linked post is only updated one way and not the other.

I can't be sure without checking more thoroughly, but my guess is that there's a nesting of the `save_post` filter going on, which therefore falls victim to https://core.trac.wordpress.org/ticket/17817 whereby subsequent filters do not fire on the way out of the nesting.

**Comment by dnaber-de** [comment]
Thanks for the hint, Christian. 

For me the problem appears only on the current version (2.1.2) from the pro repository. When deactivating it and activate the master of this free repo, everything works (almost) like expected:

```
[Fri Mar 20 16:52:05.715899 2015] [:error] [pid 19831] [client 127.0.0.1:33799] PHP Notice:  save_post_callback_10/blog/1 in /var/www/projects/wp-mlp/wp-content/plugins/mlp-debug/mlp-debug.php on line 29
[Fri Mar 20 16:52:05.715952 2015] [:error] [pid 19831] [client 127.0.0.1:33799] PHP Notice:  save_post_callback_11/blog/1 in /var/www/projects/wp-mlp/wp-content/plugins/mlp-debug/mlp-debug.php on line 23
[Fri Mar 20 16:52:05.716828 2015] [:error] [pid 19831] [client 127.0.0.1:33799] PHP Notice:  mlp_before_post_synchronization_callback_10/blog/1 in /var/www/projects/wp-mlp/wp-content/plugins/mlp-debug/mlp-debug.php on line 35
[Fri Mar 20 16:52:05.733175 2015] [:error] [pid 19831] [client 127.0.0.1:33799] PHP Notice:  save_post_callback_10/blog/2 in /var/www/projects/wp-mlp/wp-content/plugins/mlp-debug/mlp-debug.php on line 29
[Fri Mar 20 16:52:05.733234 2015] [:error] [pid 19831] [client 127.0.0.1:33799] PHP Notice:  save_post_callback_11/blog/2 in /var/www/projects/wp-mlp/wp-content/plugins/mlp-debug/mlp-debug.php on line 23
[Fri Mar 20 16:52:05.734788 2015] [:error] [pid 19831] [client 127.0.0.1:33799] PHP Notice:  mlp_after_post_synchronization_callback_10/blog/1 in /var/www/projects/wp-mlp/wp-content/plugins/mlp-debug/mlp-debug.php on line 41
[Fri Mar 20 16:52:05.737217 2015] [:error] [pid 19831] [client 127.0.0.1:33799] PHP Notice:  save_post_callback_10/blog/1 in /var/www/projects/wp-mlp/wp-content/plugins/mlp-debug/mlp-debug.php on line 29
```

**Comment by christianwach** [comment]
Hmm, I thought they were the same since ceb788cd98c935bcacfa326eee1e38f78c297874. Probably needs a dev to chime and explain the difference.

**Comment by toscho** [comment]
There is no relevant difference. At least, I’m not aware of one.

**Comment by dnaber-de** [comment]
I think the last test was overhasty. Maybe a draft or non translated post. However, it still does not work, even with the free repo.

Guess I'm dealing with the problem, described here in the second case: http://xplus3.net/2011/08/18/wordpress-action-nesting/

**Comment by christianwach** [comment]
@dnaber-de Yep, that's the issue https://core.trac.wordpress.org/ticket/17817 is trying to solve. One possible way to work around this is to call `add_action('some_delayed_action')` during the `save_post` action. Then, after `save_post` has completed, call `do_action('some_delayed_action')` and call `wp_insert_post()` at that stage.

**Comment by christianwach** [comment]
@toscho Thanks for clarfying.

Any thoughts on this issue?

**Comment by dnaber-de** [comment]
I'm not sure if it is an issue of MLP. 

Here's the problem in a condensed yet isolated form:

```php
<?php
namespace FilterPriority;

add_action( 'wp_loaded', __NAMESPACE__ . '\init' );

function init() {

	add_action( 'something', __NAMESPACE__ . '\call_do_something' );
	add_action( 'something', __NAMESPACE__ . '\detector', 10 );
	add_action( 'something', __NAMESPACE__ . '\detector', 11 );

	do_something();
}

function do_something() {

	return do_action( 'something' );
}

function call_do_something() {

	remove_filter( current_filter() , __FUNCTION__ );
	return do_something();
}

function detector() {
	global $wp_filter;

	$file     = WP_CONTENT_DIR . '/debug.log';
	$dt = new \DateTime( 'now' );
	$time = $dt->format( 'Y-m-d H:i:s' );
	$pid  = getmypid();
	$hook     = current_filter();
	$priority = key( $wp_filter[ $hook ] );
	$site     = get_current_blog_id();
	$msg      = "[$time] [$pid] $hook/$priority/$site\n";

	file_put_contents( $file, $msg, FILE_APPEND | LOCK_EX );
}
```

From the logs:
```
[2015-03-20 17:11:16] [19160] something/10/1
[2015-03-20 17:11:16] [19160] something/11/1
[2015-03-20 17:11:16] [19160] something//1
```

Conclusion: In my situation I need to hook into `save_post` after all typical meta box controller has stored their values. So I chose `11` as priority. I'll try to place my callback as late as possible with priority `10` so that it gets called after all the other callbacks.

**Comment by christianwach** [comment]
Yes, you're describing the nested `do_action()` bug in WordPress. But, as you can see, no amount of priority-jiggling will solve the issue, because the "outer" action will never be called.

**Comment by dnaber-de** [comment]
Maybe we should consider to move to the `wp_insert_post` action instead of `save_post` because `save_post` is much more common out there.

**Comment by kraftner** [comment]
Moving to `wp_insert_post` sound like a possible workaround. 

But there also has been some movement in [17817](https://core.trac.wordpress.org/ticket/17817) recently. Maybe this is even solved in core soonish...

**Comment by tfrommen** [comment]
Closing this now, as WordPress 4.7 resolved this. Finally. Somehow.
